### PR TITLE
HPCC-31857 : Develop an automated ECL Watch Test Suite

### PR DESCRIPTION
**Pull request by Nisha-Bagdwal**
Developed a testing framework using Java, Selenium, and TestNG. The framework is initiated by `TestRunner.java`, which loads all test classes listed in `config/TestClasses.java` and executes them sequentially. Each test class contains at least one method annotated with `@Test`, which serves as the entry point for the tests. Test cases cover the Activities and ECL Workunit pages, including tests for text presence, link functionality, sorting order, workunits content, description, and checkbox functionality. Added loggers in separate files for error, debug, and detail. Updated YML files to reflect jars and commands for the current testing framework. Additionally, provided comprehensive documentation with UML diagrams and detailed explanations of each method within the classes.

Signed-off-by: Nisha Bagdwal <nishabagdwal@gmail.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
I tested the code locally manually.

### CODE REVIEW

**Review by cloLN — CHANGES_REQUESTED** [review]
I would prefer if your directories were more generalized.  just replace your username with something like yourusername.

**Review by AttilaVamos — COMMENTED** [review]
Some questions.

File: esp/src/test-ui/tests/framework/README.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+- "detail" means generate error and exception log file with a detailed debug file.
+- If no -l and log_level is passed in the argument, only error and exception log will be generated
+- "path" is the path of the folder where the json files are
+- The code will log an error if the '-p' and 'path' arguments are not provided, as the JSON folder path is required for the test suite.
+
+path could be something like:
+
+for GitHub Actions -> /home/runner/HPCCSystems-regression/log/
+
+for local machine -> C:/Users/nisha/Documents/Internship/Work/files/
```
[the remark this answers is not in the document]
Comment (Nisha-Bagdwal) [reply esp/src/test-ui/tests/framework/README.md]: Yes, I agree with you. I have made the change.

File: esp/src/test-ui/tests/framework/config/Config.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
+package framework.config;
+
+public class Config {
+
+    public static final String LOG_FILE_ERROR = "error_ecl_test.log";
+    public static final String LOG_FILE_EXCEPTION = "exception_ecl_test.log";
+    public static final String LOG_FILE_DEBUG = "debug_ecl_test.log";
+    public static final String LOG_FILE_DETAIL = "detail_ecl_test.log";
+    public static final String LOCAL_OS = "Windows";
+    public static final String LOCAL_USER_PROFILE = "C:\\Users\\nisha";
```
[the remark this answers is not in the document]
Comment (Nisha-Bagdwal) [reply esp/src/test-ui/tests/framework/config/Config.java]: Done

File: esp/src/test-ui/tests/framework/config/Config.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    public static final String PATH_LOCAL_CHROME_DRIVER = "C:/Users/nisha/Documents/Internship/Work/jars/chromeDriver/chromedriver.exe";
+    public static final String PATH_GH_ACTION_CHROME_DRIVER = "/usr/bin/chromedriver";
+    public static final int[] dropdownValues = {10, 25, 50, 100, 250, 500, 1000};
```
[the remark this answers is not in the document]
Comment (Nisha-Bagdwal) [reply esp/src/test-ui/tests/framework/config/Config.java:13]: Yes, this array is used in the logic where I have to select the smallest dropdown value that is greater than the total number of items in JSON.

File: esp/src/test-ui/tests/framework/pages/ActivitiesTest.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+import static framework.config.URLConfig.urlMap;
+
+public class ActivitiesTest {
+
+    static final String[] textArray = {"Target/Wuid", "Graph", "State", "Owner", "Job Name"};
+
+    @Test
+    public void testActivitiesPage() {
+        if(!Common.openWebPage(urlMap.get(URLConfig.NAV_ACTIVITIES).getUrl())){
+            return;
```
[the remark this answers is not in the document]
Comment (Nisha-Bagdwal) [reply esp/src/test-ui/tests/framework/pages/ActivitiesTest.java:24]: Yes, that code is written in the Common.openWebPage method. Inside that method, the code tries to open the web page using the URL and if any exception occurs during that time, the code catches that exception and logs it in the error logs along with the URL it tried to open.

File: esp/src/test-ui/tests/framework/documentation/DesignDocument.md [hunk 5]
```diff
@@ -0,0 +1,1249 @@
...
+### Java Classes for Representing Logical Files JSON Data
+
+This section details the class structure used to map JSON data file of list of "Logical Files"
+entities into Java objects. These classes provide a clear representation of the data and
+allow for easy access to its values throughout the codebase. This structure is particularly
+beneficial for writing test cases, as it simplifies working with the data in a well-defined
+format. Below are the UML diagrams of the classes used for JSON mapping to java objects for
+logical files JSON files.
+
+![img.png](DFULogicalFile.png)
```
[the remark this answers is not in the document]
Comment (Nisha-Bagdwal) [reply esp/src/test-ui/tests/framework/documentation/DesignDocument.md:1103]: Added the explanation for both DFULogicalFiles and Workunits JSON

File: esp/src/test-ui/tests/framework/config/Config.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    public static final int MALFORMED_TIME_STRING = -1;
+    public static final int WAIT_TIME_IN_SECONDS = 1;
+    public static final int WAIT_TIME_THRESHOLD_IN_SECONDS = 20;
+    public static final String TEST_DESCRIPTION_TEXT = "Testing Description";
+    public static final boolean TEST_DETAIL_PAGE_FIELD_NAMES_ALL = true;
```
[the remark this answers is not in the document]
Comment (Nisha-Bagdwal) [reply esp/src/test-ui/tests/framework/config/Config.java]: Done

File: esp/src/test-ui/tests/framework/TestRunner.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        }
+
+        if (log_level != null) {
+            Config.LOG_LEVEL = log_level;
+        }
+
+        if (path != null) {
+            Config.PATH_FOLDER_JSON = path;
+        } else {
+            Common.logError("Error: JSON folder path is required. Use -p <path> to specify the path.");
```
[the remark this answers is not in the document]
Comment (Nisha-Bagdwal) [reply esp/src/test-ui/tests/framework/TestRunner.java:66]: done

File: esp/src/test-ui/tests/framework/TestRunner.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    // Parses the command-line arguments to set the log level and JSON folder path.
+    // The method checks for the presence of '-l' (log_level) and '-p' (path) arguments.
+    // Logs an error if the '-p' argument is not provided, as the JSON path is required.
+    // path is the path of the folder where the json files are
+    // log level is of two types "debug" and "detail"
+    // "debug" means generate error log file with a debug log file.
+    // "detail" means generate error log file with a detailed debug file.
+    // if no -l and log level is passed in the argument, only error log will be generated
+
+    public static void getLogLevelAndJSONFolderPath(String[] args) { // -l <log_level> -p <path>
```
[the remark this answers is not in the document]
Comment (Nisha-Bagdwal) [reply esp/src/test-ui/tests/framework/TestRunner.java]: Done

**Review by cloLN — APPROVED** [review]
changes look good to me.

**Review by AttilaVamos — CHANGES_REQUESTED** [review]
Only a couple of small comments, questions. 
If any change has larger impact than modify the code in the current file, please discuss it with me before change.

File: esp/src/test-ui/tests/Activities.java [hunk 9]
```diff
@@ -24,7 +24,7 @@ public static void main(String[] args) throws IOException, InterruptedException
         Capabilities caps = ((RemoteWebDriver) driver).getCapabilities();
 
         String browserName = caps.getBrowserName();
-        String browserVersion = caps.getVersion();
+        //String browserVersion = caps.getVersion();
```
[the remark this answers is not in the document]
Comment (Nisha-Bagdwal) [reply esp/src/test-ui/tests/Activities.java:27]: Not applicable for my test developement

File: esp/src/test-ui/tests/framework/model/NavigationWebElement.java [hunk 10]
```diff
@@ -0,0 +1,14 @@
+package framework.model;
+
+import org.openqa.selenium.WebElement;
+
+public record NavigationWebElement(String name, String hrefValue, WebElement webElement) {
```
[the remark this answers is not in the document]
Comment (Nisha-Bagdwal) [reply esp/src/test-ui/tests/framework/model/NavigationWebElement.java:5]: Yes, we need it

File: esp/src/test-ui/tests/framework/pages/BaseTableTest.java [hunk 11]
```diff
@@ -0,0 +1,591 @@
...
+            testForAllText();
+
+            jsonObjects = getAllObjectsFromJson();
+            if (jsonObjects != null) {
+
+                int numOfItemsJSON = jsonObjects.size();
+                clickDropdown(numOfItemsJSON);
+
+                testContentAndSortingOrder();
+            }
```
[the remark this answers is not in the document]
Comment (Nisha-Bagdwal) [reply esp/src/test-ui/tests/framework/pages/BaseTableTest.java:90]: Problem is handled in getAllObjectsFromJson method

File: esp/src/test-ui/tests/framework/SetupForDev.md [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+    - Visit the Oracle VirtualBox [website](https://www.virtualbox.org/) and download the latest version for Windows.
+    - Follow the installation instructions to install VirtualBox on your system.
+
+2. **Download the Ubuntu 22.04 Desktop ISO file:**
+    - Go to the Ubuntu [downloads page](https://ubuntu.com/download/desktop) and download the Ubuntu 22.04 LTS ISO file.
+
+3. **Set up a new VM in VirtualBox:**
+    - Open VirtualBox and click on `New` to create a new virtual machine.
+    - Name the VM and select the type and version (Linux, Ubuntu 64-bit).
+    - Configure system settings such as memory size and hard disk (create a virtual hard disk now).
```
[the remark this answers is not in the document]
Comment (Nisha-Bagdwal) [reply esp/src/test-ui/tests/framework/SetupForDev.md:17]: Done

File: esp/src/test-ui/tests/framework/SetupForDev.md [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+   ```sh
+   sudo dpkg -l | grep 'hpccsystems-pl'
+   ```
+5. **Start HPCC-Platform:**
+   ```sh
+   sudo /etc/init.d/hpcc-init start
+   ```
+
+6. **Verify access to ECL Watch:**
+    - Open a browser on your local machine and go to `http://192.168.0.221:8010/` to check if you can access ECL Watch.
```
[the remark this answers is not in the document]
Comment (Nisha-Bagdwal) [reply esp/src/test-ui/tests/framework/SetupForDev.md]: Done

File: esp/src/test-ui/tests/framework/model/DFULogicalFile.java [hunk 14]
```diff
@@ -0,0 +1,382 @@
...
+    }
+
+    public int getMaxSkewPart() {
+        return maxSkewPart;
+    }
+
+    public void setMaxSkewPart(int maxSkewPart) {
+        this.maxSkewPart = maxSkewPart;
+    }
+}
```
[the remark this answers is not in the document]
Comment (Nisha-Bagdwal) [reply esp/src/test-ui/tests/framework/model/DFULogicalFile.java]: Done

File: esp/src/test-ui/tests/framework/pages/ECLWorkUnitsTest.java [hunk 15]
```diff
@@ -0,0 +1,424 @@
...
+            WebElement element = Common.waitForElement(By.id("wuid"));
+            if (element != null) {
+                return element.getAttribute("title");
+            }
+        } catch (Exception ex) {
+            Common.logException("Error: " + getPageName() + ex.getMessage(), ex);
+        }
+        return "Invalid Page";
+    }
+}
```
[the remark this answers is not in the document]
Comment (Nisha-Bagdwal) [reply esp/src/test-ui/tests/framework/pages/ECLWorkUnitsTest.java]: Done

File: esp/src/test-ui/tests/framework/pages/FilesLogicalFilesTest.java [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+            case "descending" -> descendingSortJson(columnKey);
+            case "none" -> descendingSortJson(sortByColumnKeyWhenSortedByNone);
+        }
+    }
+
+    @Override
+    protected String getCurrentPage() {
+        return "Invalid Page";
+    }
+}
```
[the remark this answers is not in the document]
Comment (Nisha-Bagdwal) [reply esp/src/test-ui/tests/framework/pages/FilesLogicalFilesTest.java]: Done

File: esp/src/test-ui/tests/framework/pages/BaseTableTest.java [hunk 17]
```diff
@@ -0,0 +1,591 @@
...
+        try {
+
+            WebElement dropdown = Common.waitForElement(By.id("pageSize"));
+            dropdown.click();
+
+            WebDriverWait wait = new WebDriverWait(Common.driver, Duration.ofSeconds(Config.WAIT_TIME_THRESHOLD_IN_SECONDS));
+
+            WebElement dropdownList = wait.until(ExpectedConditions.visibilityOfElementLocated(By.id("pageSize-list")));
+
+            List<WebElement> options = dropdownList.findElements(By.tagName("button"));
```
[the remark this answers is not in the document]
Comment (Nisha-Bagdwal) [reply esp/src/test-ui/tests/framework/pages/BaseTableTest.java:539]: I have made the change in the code and removed the hardcoded Config.dropdownValues

File: esp/src/test-ui/tests/framework/pages/BaseTableTest.java [hunk 18]
```diff
@@ -0,0 +1,591 @@
...
+            // the smallest dropdown value greater than numOfItemsJSON
+            for (int value : Config.dropdownValues) {
+                if (numOfItemsJSON < value) {
+                    selectedValue = value;
+                    break;
+                }
+            }
+
+            for (WebElement option : options) {
+                if (option.getText().equals(String.valueOf(selectedValue))) {
```
[the remark this answers is not in the document]
Comment (Nisha-Bagdwal) [reply esp/src/test-ui/tests/framework/pages/BaseTableTest.java]: Hardcoded Config.dropdownValues has been removed now

**Review by AttilaVamos — APPROVED** [review]
Seems good. 
Please squash it.

**Comment by Nisha-Bagdwal** [comment]
@ghalliday Please merge my pull request.

**Comment by AttilaVamos** [comment]
@JamesDeFabia Can you please take a look for the documentations:
- esp/src/test-ui/tests/framework/README.md
- esp/src/test-ui/tests/framework/SetupForDev.md
- esp/src/test-ui/tests/framework/documentation/DesignDocument.md

File: esp/src/test-ui/tests/framework/SetupForDev.md [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+    - Link the ISO file by going to the `Settings` of the VM, navigating to `Storage`, and attaching the ISO file to the optical drive.
+    - Boot the VM and follow the installation wizard to complete the Ubuntu setup.
+
+#### Installing HPCC-Platform on the VM
+
+**After successfully installing the VM, proceed with installing the HPCC-Platform:**
+
+1. **Download the HPCC-Platform package:**
+   ```sh
+   wget https://cdn.hpccsystems.com/releases/CE-Candidate-9.8.2/bin/platform/hpccsystems-platform-community_9.8.2-1jammy_amd64_withsymbols.deb
```
Comment (JamesDeFabia) [inline esp/src/test-ui/tests/framework/SetupForDev.md:29]: This should show how to get the latest release and use a version of 9.8.x as an example

File: esp/src/test-ui/tests/framework/SetupForDev.md [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+   ```
+
+2. **Install the package:**
+   ```sh
```
Comment (JamesDeFabia) [inline esp/src/test-ui/tests/framework/SetupForDev.md:33]: This should show how to install the package downloaded and use a version of 9.8.x as an example

**Review by JamesDeFabia — CHANGES_REQUESTED** [review]
Only a couple of minor comments inline.

File: esp/src/test-ui/tests/framework/SetupForDev.md (hunk 19 above)
Comment (Nisha-Bagdwal) [reply esp/src/test-ui/tests/framework/SetupForDev.md]: Done

File: esp/src/test-ui/tests/framework/SetupForDev.md (hunk 20 above)
Comment (Nisha-Bagdwal) [reply esp/src/test-ui/tests/framework/SetupForDev.md:45]: Done.

**Review by JamesDeFabia — APPROVED** [review]
Looks good from my POV

**Pull request merged**